### ISSUE TITLE
Browser crash by scaling after geoCenter

### ISSUE THREAD

**Issue by nyurik** [vega/vega#978]
Repo:
* Copy this code to the Vega editor, observe that the the map is being shown ok.
* Add   `, {"type": "formula", "expr": "geoCentroid(datum)", "as": "center"}`  as the last transformation after "filter" in "countries"
* Observe full browser crash
* Try the same, but first delete the "scaling" section before pasting the geoCenter transformation - observe that it works fine

```json
{
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "signals": [
    {"name": "zoom", "value": 1},
    {"name": "latitude", "value": 0},
    {"name": "longitude", "value": 0}
  ],
  "data": [
    {
      "name": "logCount",
      "values": [
        {"key": "US", "doc_count": 100},
        {"key": "FR", "doc_count": 1}
      ]
    },
    {
      "name": "countries",
      "url": "https://layers.geo.elastic.co/blob/5659313586569216?elastic_tile_service_tos=agree",
      "format": {"type": "json", "property": "features"},
      "transform": [
        {
          "type": "lookup",
          "from": "logCount",
          "key": "key",
          "fields": ["properties.iso2"],
          "values": ["doc_count"]
        },
        {"type": "filter", "expr": "datum.doc_count != null"}
      ]
    }
  ],
  "scales": [
    {
      "name": "color",
      "domain": {"data": "countries", "field": "doc_count"},
      "range": {"scheme": "Greens"}
    }
  ],
  "marks": [
    {
      "type": "shape",
      "from": {"data": "countries"},
      "encode": {"enter": {"fill": {"value": "#000"}}},
      "transform": [
        {"type": "geoshape", "projection": "projection"}
      ]
    }
  ],
  "projections": [
    {
      "name": "projection",
      "type": "mercator",
      "scale": {"signal": "256*pow(2,zoom)/2/PI"},
      "rotate": [{"signal": "-longitude"}, 0, 0],
      "center": [0, {"signal": "latitude"}],
      "translate": [
        {"signal": "width/2"},
        {"signal": "height/2"}
      ]
    }
  ]
}
```

**Comment by jheer** [comment]
[geoCentroid](https://vega.github.io/vega/docs/expressions/#geoCentroid) requires two arguments: a projection, then the GeoJSON feature. If you add a projection as the first argument the problem goes away.

Of course, an error / warning message from the geo method could be helpful here...

**Comment by nyurik** [comment]
@jheer thanks, but it is strange that Chrome did a full crash... I wonder what caused it - seemed like a infinite loop?  Note that rendering works fine if scaling is removed.

**Comment by jheer** [comment]
Yes, I believe it caused an infinite loop.

**Comment by nyurik** [comment]
@jheer sounds like a security issue with the user contributed code if run server-side

**Comment by jheer** [comment]
The infinite loop was due to a different issue triggered by this issue: improper error handling when a cycle is present in the dataflow graph. This has been fixed in vega-dataflow 3.0.2.
